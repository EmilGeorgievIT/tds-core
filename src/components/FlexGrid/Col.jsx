import React from 'react'
import PropTypes from 'prop-types'
import { Subscriber } from 'react-broadcast'

import { Col as ReactFlexboxGridCol } from 'react-flexbox-grid'

import safeRest from '../../utils/safeRest'

const removeProps = ({
  xs,
  sm,
  md,
  lg,
  xl,
  xsOffset,
  smOffset,
  mdOffset,
  lgOffset,
  xlOffset,
  ...rest
}) => safeRest(rest)

const Col = ({ span, offset, children, ...rest }) => (
  <Subscriber channel="gutterless">
    {gutterless => {
      return (
        <ReactFlexboxGridCol
          {...removeProps(rest)}
          xs={span || true}
          xsOffset={offset}
          style={gutterless}
        >
          {children}
        </ReactFlexboxGridCol>
      )
    }}
  </Subscriber>
)

Col.propTypes = {
  span: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
  offset: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]),
  children: PropTypes.node.isRequired,
}

Col.defaultProps = {
  span: undefined,
  offset: undefined,
}

export default Col
