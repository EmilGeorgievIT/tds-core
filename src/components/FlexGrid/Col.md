Columns automatically expand to fill the width of its container. You can specify a span amount from 1 to 12.
