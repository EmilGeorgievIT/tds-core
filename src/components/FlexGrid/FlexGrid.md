The `FlexGrid` component is a wrapper over [react-flexbox-grid](https://github.com/roylee0704/react-flexbox-grid). 

### Basic usage

FlexGrid is handy for basic grid layouts or 1-dimensional positioning. It is used
in conjunction with the [Col](#col) and [Row](#row) components.

```jsx
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col>half</FlexGrid.Col>
    <FlexGrid.Col>half</FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

### Responsive example

```jsx
<FlexGrid>
  <Responsive minWidth="sm">
  {
    matches => {
      var desktopSpan = matches ? undefined : 12
      var text = matches ? 'half width on desktop' : 'full width on mobile'

      return (
        <FlexGrid.Row>
          <FlexGrid.Col span={desktopSpan}>{text}</FlexGrid.Col>
          <FlexGrid.Col span={desktopSpan}>{text}</FlexGrid.Col>
        </FlexGrid.Row>
      )
    }
  }
  </Responsive>
</FlexGrid>
```
