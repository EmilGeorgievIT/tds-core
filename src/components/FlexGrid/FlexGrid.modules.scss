@import '../../scss/settings/responsive-grid';
@import '../../scss/utility/responsive';

$max-widths: (xs: auto, small: 544px, medium: 736px, large: 960px, xl: 1168px);

.flexGrid {
  display: flex;
  flex-flow: row wrap;
  padding: 0;
}

.centre {
  margin: 0 auto;
}

/*
  Possible alternatives to responsive CSS
  1) use mutiple 'react-media' queries when it becomes available
  https://github.com/ReactTraining/react-media/issues/69
  2) use multiple Responsive components nested in React 16.2 fragments once all teams
  adopt it https://reactjs.org/blog/2017/11/28/react-v16.2.0-fragment-support.html
*/
.limitWidth {
  @each $viewport, $max-width in $max-widths {
    @include from-breakpoint($viewport) {
      max-width: $max-width;
      margin: 0 auto;
    }
  }
}

.gutterless {
  padding-left: 0 !important;
  padding-right: 0 !important;
}

.flexGridColumn {
  padding-left: 1rem;
  padding-right: 1rem;
}

// Row styles
.flexGridRow {
  margin: 0;
  flex: 1 1 auto;
}
