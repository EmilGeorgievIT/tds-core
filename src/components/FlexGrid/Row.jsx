import React from 'react'
import PropTypes from 'prop-types'
import { Subscriber } from 'react-broadcast'

import { Row as ReactFlexboxGridRow } from 'react-flexbox-grid'

import safeRest from '../../utils/safeRest'

const Row = ({ children, ...rest }) => (
  <Subscriber channel="flex-grid">
    {gutterStyle => (
      <ReactFlexboxGridRow {...safeRest(rest)} style={gutterStyle}>
        {children}
      </ReactFlexboxGridRow>
    )}
  </Subscriber>
)

Row.propTypes = {
  children: PropTypes.node.isRequired,
}

export default Row
