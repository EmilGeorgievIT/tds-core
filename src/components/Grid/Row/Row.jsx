import React from 'react';
import PropTypes from 'prop-types';

/**
 * Align grid content horizontally.
 *
 * _This component can only be used as a child of `Grid.Container`, and must be accessed as a
 * name-spaced component: `Grid.Row`._
 *
 * @see See [Grid.Column](#column) for grid usage.
 */
function Row(props) {
  const { className, ...extraProps } = props;
  const classes = ['grid-row', className];

  return (
    <div className={classes.filter(c => c).join(' ')} {...extraProps}>
      {props.children}
    </div>
  );
}

Row.propTypes = {
  /**
   * One or more CSS class names separated by spaces to append onto the container.
   * Don't advertise as we may remove this feature soon.
   *
   * @ignore
   */
  className: PropTypes.string,
  /**
   * The grid's columns. Must be TDS `Grid.Column` components.
   *
   * @see See [Grid.Column](#column)
   */
  children: PropTypes.node.isRequired
};

export default Row;
