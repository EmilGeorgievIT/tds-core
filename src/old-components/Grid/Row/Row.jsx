import React from 'react'
import PropTypes from 'prop-types'

/**
 * <span class="docs--badge__deprecated">deprecated</span>
 *
 * <strong>The grid components are deprecated and should not be used. Use the [FlexGrid](#flexgrid) component instead.</strong>
 *
 * Align grid content horizontally.
 *
 * _This component can only be used as a child of `Grid.Container`, and must be accessed as a
 * name-spaced component: `Grid.Row`._
 *
 * @see See [Grid.Column](#column) for grid usage.
 */
function Row(props) {
  const { className, ...extraProps } = props
  const classes = ['grid-row', className]

  return (
    <div className={classes.filter(c => c).join(' ')} {...extraProps}>
      {props.children}
    </div>
  )
}

Row.propTypes = {
  /**
   * One or more CSS class names separated by spaces to append onto the container.
   * Don't advertise as we may remove this feature soon.
   *
   * @ignore
   */
  className: PropTypes.string, // eslint-disable-line react/require-default-props
  /**
   * The grid's columns. Must be TDS `Grid.Column` components.
   *
   * @see See [Grid.Column](#column)
   */
  children: PropTypes.node.isRequired,
}

export default Row
