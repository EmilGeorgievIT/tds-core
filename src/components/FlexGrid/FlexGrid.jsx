import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast } from 'react-broadcast'

import { Grid } from 'react-flexbox-grid'

import safeRest from '../../utils/safeRest'

import Col from './Col'
import Row from './Row'

const FlexGrid = ({ gutterless, children, ...rest }) => {
  const css = gutterless ? { padding: 0, margin: 0 } : undefined

  return (
    <Broadcast channel="gutterless" value={css}>
      <Grid {...safeRest(rest)} fluid style={css}>
        {children}
      </Grid>
    </Broadcast>
  )
}

FlexGrid.propTypes = {
  gutterless: PropTypes.bool,
  children: PropTypes.node.isRequired,
}

FlexGrid.defaultProps = {
  gutterless: false,
}

FlexGrid.Row = Row
FlexGrid.Col = Col

export default FlexGrid
