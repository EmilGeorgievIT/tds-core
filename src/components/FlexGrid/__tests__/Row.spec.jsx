import React from 'react'
import { mount } from 'enzyme'
import { Row as ReactFlexboxGridRow } from 'react-flexbox-grid'

import FlexGrid from '../FlexGrid'
import Row from '../Row'

describe('Col', () => {
  const doShallow = (props = {}) => {
    const wrapper = mount(
      <FlexGrid>
        <Row {...props}>Some content</Row>)
      </FlexGrid>
    )
    return wrapper.find(ReactFlexboxGridRow)
  }

  it('renders', () => {
    const col = doShallow()
    expect(col).toMatchSnapshot()
  })

  it('renders a react-flexbox-grid Col component', () => {
    const row = doShallow()

    expect(row).toMatchSelector(ReactFlexboxGridRow)
  })

  it('passes additional attributes to the react-flexbox-grid Col', () => {
    const col = doShallow({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(col).toHaveProp('id', 'the-id')
    expect(col).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const col = doShallow({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(col).not.toHaveProp('className', 'my-custom-class')
    expect(col).not.toHaveProp('style', { color: 'hotpink' })
  })
})
