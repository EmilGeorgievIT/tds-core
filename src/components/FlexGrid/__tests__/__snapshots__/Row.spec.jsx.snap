// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col renders 1`] = `
<Row>
  <div
    className="row"
  >
    Some content
  </div>
</Row>
`;
